lib/ts/bwprint.cc

```diff
@@ -0,0 +1,410 @@
+// in decimal.
+inline int
+tv_to_positive_decimal(ts::TextView src, ts::TextView *out)
+{
+  int zret = 0;
+
+  if (out) {
+    out->clear();
+  }
+  if (src.ltrim_if(&isspace) && src) {
+    const char *start = src.data();
+    const char *limit = start + src.size();
+    while (start < limit && ('0' <= *start && *start <= '9')) {
+      zret = zret * 10 + *start - '0';
+      ++start;
+    }
+    if (out && (start > src.data())) {
+      out->set_view(src.data(), start);
+    }
+  }
+  return zret;
+}
+}
+
+/// Parse a format specification.
+ts::BWFSpec::BWFSpec(TextView fmt)
+{
+  TextView num;
+  intmax_t n;
+
+  _name = fmt.take_prefix_at(':');
+  // if it's parsable as a number, treat it as an index.
+  n = tv_to_positive_decimal(_name, &num);
+  if (num)
+    _idx = static_cast<decltype(_idx)>(n);
+
+  if (fmt) {
+    TextView sz = fmt.take_prefix_at(':'); // the format specifier.
+    _ext        = fmt;                     // anything past the second ':' is the extension.
+    if (sz) {
+      // fill and alignment
+      if ('%' == *sz) { // enable URI encoding of the fill character so metasyntactic chars can be used if needed.
+        if (sz.size() < 4) {
+          throw std::invalid_argument("Fill URI encoding without 2 hex characters and align mark");
+        }
+        if (Align::NONE == (_align = align_of(sz[3]))) {
+          throw std::invalid_argument("Fill URI without alignment mark");
+        }
+        char d1 = sz[1], d0 = sz[2];
+        if (!isxdigit(d0) || !isxdigit(d1)) {
+          throw std::invalid_argument("URI encoding with non-hex characters");
+        }
+        _fill = isdigit(d0) ? d0 - '0' : tolower(d0) - 'a' + 10;
+        _fill += (isdigit(d1) ? d1 - '0' : tolower(d1) - 'a' + 10) << 4;
+        sz += 4;
+      } else if (sz.size() > 1 && Align::NONE != (_align = align_of(sz[1]))) {
+        _fill = *sz;
+        sz += 2;
+      } else if (Align::NONE != (_align = align_of(*sz))) {
+        ++sz;
+      }
+      if (!sz)
+        return;
+      // sign
+      if (is_sign(*sz)) {
+        _sign = *sz;
+        if (!++sz)
+          return;
+      }
+      // radix prefix
+      if ('#' == *sz) {
+        _radix_lead_p = true;
+        if (!++sz)
+          return;
+      }
+      // 0 fill for integers
+      if ('0' == *sz) {
+        if (Align::NONE == _align)
+          _align = Align::SIGN;
+        _fill    = '0';
+        ++sz;
+      }
+      n = tv_to_positive_decimal(sz, &num);
+      if (num) {
+        _min = static_cast<decltype(_min)>(n);
+        sz.remove_prefix(num.size());
+        if (!sz)
+          return;
+      }
+      // precision
+      if ('.' == *sz) {
+        n = tv_to_positive_decimal(++sz, &num);
+        if (num) {
+          _prec = static_cast<decltype(_prec)>(n);
+          sz.remove_prefix(num.size());
+          if (!sz)
+            return;
+        } else {
+          throw std::invalid_argument("Precision mark without precision");
```

Comment:
Generic use of exceptions is questionable in a source base that has not be upgraded to exception safety.